src/MessagePack.UnityClient/Assets/Scripts/MessagePack/MessagePackSerializer.cs

```diff
@@ -470,50 +473,99 @@ private static int LZ4Operation(in ReadOnlySequence<byte> input, Span<byte> outp
+                        var extReader = new MessagePackReader(extension.Data);
+
+                        // The first part of the extension payload is a MessagePack-encoded Int32 that
+                        // tells us the length the data will be AFTER decompression.
+                        int uncompressedLength = extReader.ReadInt32();
+
+                        // The rest of the payload is the compressed data itself.
+                        ReadOnlySequence<byte> compressedData = extReader.Sequence.Slice(extReader.Position);
+
+                        Span<byte> uncompressedSpan = writer.GetSpan(uncompressedLength).Slice(0, uncompressedLength);
+                        int actualUncompressedLength = LZ4Operation(compressedData, uncompressedSpan, LZ4CodecDecode);
+                        Debug.Assert(actualUncompressedLength == uncompressedLength, "Unexpected length of uncompressed data.");
+                        writer.Advance(actualUncompressedLength);
+                        return true;
+                    }
+                }
+
+                // Try to find LZ4BlockArray
+                if (reader.NextMessagePackType == MessagePackType.Array)
                 {
-                    // Read the extension using the original reader, so we "consume" it.
-                    ExtensionResult extension = reader.ReadExtensionFormat();
-                    var extReader = new MessagePackReader(extension.Data);
-
-                    // The first part of the extension payload is a MessagePack-encoded Int32 that
-                    // tells us the length the data will be AFTER decompression.
-                    int uncompressedLength = extReader.ReadInt32();
-
-                    // The rest of the payload is the compressed data itself.
-                    ReadOnlySequence<byte> compressedData = extReader.Sequence.Slice(extReader.Position);
-
-                    Span<byte> uncompressedSpan = writer.GetSpan(uncompressedLength).Slice(0, uncompressedLength);
-                    int actualUncompressedLength = LZ4Operation(compressedData, uncompressedSpan, LZ4Codec.Decode);
-                    Debug.Assert(actualUncompressedLength == uncompressedLength, "Unexpected length of uncompressed data.");
-                    writer.Advance(actualUncompressedLength);
-                    return true;
+                    MessagePackReader peekReader = reader.CreatePeekReader();
+                    var arrayLength = peekReader.ReadArrayHeader();
+                    if (arrayLength != 0 && peekReader.NextMessagePackType == MessagePackType.Extension)
+                    {
+                        ExtensionHeader header = peekReader.ReadExtensionFormatHeader();
+                        if (header.TypeCode == ThisLibraryExtensionTypeCodes.Lz4BlockArray)
+                        {
+                            // switch peekReader as original reader.
+                            reader = peekReader;
```

Comment:
Clever as I hadn't considered that technique, but I'm afraid it doesn't work quite right. When you call `CreatePeekReader()`, that reader is based on a slice of the original sequence (so that it starts at the right position). So when you assign the original reader to `peekReader`, you're actually significantly changing `reader.Sequence` which is a public property, and code may expect `reader.Sequence` to refer to the entire original sequence being deserialized.
However I really like your technique here, and others might try the same thing. So I think I'll send a PR to change `CreatePeekReader` so that your technique here will work properly.

Reply:
Thank you, I understand.
And thank you for the accurate review.
I expect your next PR so I keep this code.

Reply:
See #690

Reply:
See #690